src/TestSuite/Fixture/FixtureHelper.php

```diff
@@ -159,11 +171,71 @@ public function insert(array $fixtures): void
     public function truncate(array $fixtures): void
     {
         $this->runPerConnection(function (ConnectionInterface $connection, array $groupFixtures) {
-            $connection->disableConstraints(function () use ($connection, $groupFixtures) {
-                foreach ($groupFixtures as $fixture) {
+            if ($connection instanceof Connection) {
+                $sortedFixtures = $this->sortByConstraint($connection, $groupFixtures);
+            }
+
+            $driver = $connection->getDriver();
+            if (
+                isset($sortedFixtures) &&
+                (
+                    $driver instanceof Postgres ||
+                    $driver instanceof Sqlite ||
+                    $driver instanceof Sqlserver
+                )
+            ) {
+                foreach (array_reverse($sortedFixtures) as $fixture) {
                     $fixture->truncate($connection);
                 }
-            });
+            } else {
+                $connection->disableConstraints(function () use ($connection, $groupFixtures) {
+                    foreach ($groupFixtures as $fixture) {
+                        $fixture->truncate($connection);
+                    }
+                });
+            }
         }, $fixtures);
     }
+
+    /**
+     * Sort fixtures with foreign constraints last if possible, otherwise returns null.
+     *
+     * @param \Cake\Database\Connection $connection Database connection
+     * @param array<\Cake\Datasource\FixtureInterface> $fixtures Database fixtures
+     * @return array|null
+     */
+    protected function sortByConstraint(Connection $connection, array $fixtures): ?array
+    {
+        static $schemas = [];
+
+        $fixturesWithConstraint = [];
+        $fixturesWithoutConstraint = [];
+        foreach ($fixtures as $fixture) {
+            $tableName = $fixture->sourceName();
+            if (!isset($schemas[$tableName])) {
+                $schemas[$tableName] = $connection->getSchemaCollection()->describe($tableName);
+            }
+
+            $schema = $schemas[$tableName];
+            foreach ($schema->constraints() as $constraintName) {
+                $constraint = $schema->getConstraint($constraintName);
+                if (!$constraint) {
+                    continue;
+                }
+
+                if ($constraint['type'] === TableSchema::CONSTRAINT_FOREIGN) {
+                    $fixturesWithConstraint[$schema->name()] = $fixture;
+                    if (isset($fixturesWithConstraint[$constraint['references'][0]])) {
```

Comment:
Might be good to have a comment here around cross referenced fixtures causing an abort. I didn't catch that on the first read.

Reply:
I split the foreign references code into a separate function so it's easier to read.

Fixed the cross dependency check to check all references and added a comment explaining why.